examples/hashmap.dx

```diff
@@ -0,0 +1,223 @@
+' # Associative Collections
+Secretly this is actually an example on type-classes
+
+data Hashable a:Type = MkHashable (a->a->Bool) (Int->a->Int) -- isequal, hash
```

Comment:
Problem is you need `nan`isequal` nan`
So you can't use `==` which follows IEEE saying `not (nan==nan)`.

Reply:
I think it's a bit confusing to call it `isequal` then. Maybe `identical`? Because it's confusing that `==` and `isequal` are different.

Reply:
🤷  `isequal` is what julia calls it.

I tried to findout what python calls this. Turns out it doesn't and just makes a pathological dictionary.
It looks at first like it does, because it first checks object identity in terms of memory address, and then only checks `___eq___` if not the same memory address. So while usign `math.nan` everywhere works, if you ahve a `float("nan")` or a reference to `numpy.nan` anywhere then it doesn't match
https://stackoverflow.com/questions/6441857/nans-as-key-in-dictionaries
Sounds like it also silently breaks in C# https://softwareengineering.stackexchange.com/questions/308077/should-vectorfloat-equals-be-reflexive-or-should-it-follow-ieee-754-semantic
And Haskell's tree based dictionaries also: 
https://hackage.haskell.org/package/nri-prelude-0.2.0.0/docs/Basics.html#t:Ord
Hilarious.

I think some lisps like [Racket](https://docs.racket-lang.org/reference/Equality.html#%28def._%28%28quote._~23~25kernel%29._eqv~3f%29%29)
lisps seem to have multiple equality definations like that.
[Common Lisp](https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node74.html) also has a bunch but common lisp is older than IEEE Floats and only partially supports them.

Can name it what-ever you like. Names are infinately bike-shed able.
Seems like the prior art  is Julia's `isequal`, and maybe racket's `eqv?` (which could be come `equivelent`.
Could even go long and explict: `isEqualForPurposesOfUniqueness`

OTOH one could say that using Floats as keys is also open to other pathologies, since equality is rare.
But it's useful if you know and rember that you can't use it if you can't be certain to get some value without rounding.

Reply:
I wonder if there's any good reason to put floats in hash tables anyway. I can't think of one? Especially due to floating point rounding.

What if:
- `Hashable` just used `(==)` and had `Eq` as a superclass
- Floats don't implement hash and thus can't be used in maps
- Optionally, we have a newtype wrapper `HashableFloat` whose `Eq` says `nan == nan` is True

Reply:
Could do.
I think the problem is that it shows up in nested circumstances.
Like records of tables of tuples ... where 1 field is a Float.

Going through and replacing then via  optics, that could work maybe.
Seems awkward and expensive, but maybe not.


Could also abandon IEEE behavior for floats, and say `nan==nan`.

Reply:
I would lean towards @danieldjohnson's suggestion. It feels weird to use floats as keys, given all the inaccuracies associated with calculations on those. If someone really knows what they're doing then they should use a wrapper like `HashableFloat`. Otherwise it does look like a loaded footgun to me.

Reply:
done